pkg/api/v1/types.go

```diff
@@ -1217,12 +1217,14 @@ type EndpointSubset struct {
 	Ports     []EndpointPort    `json:"ports,omitempty" description:"port numbers available on the related IP addresses"`
 }
 
-// EndpointAddress is a tuple that describes single IP address.
+// EndpointAddress is a tuple that describes single endpoint address.
 type EndpointAddress struct {
-	// The IP of this endpoint.
-	// TODO: This should allow hostname or IP, see #4447.
+	// The IP of this endpoint, mutually exclusive with Hostname
 	IP string `json:"ip" description:"IP address of the endpoint"`
 
+	// Hostname for this endpoint, you can specify IP or Hostname, but not both.
+	Hostname string `json:"hostname" description:"Hostname for this endpoint"`
```

Comment:
Would we set both IP and Hostname, or just OneOf? If OneOf, then this is a breaking API change. If we overloaded the existing field, that would also be a breaking API change.

We do have both IP and Hostname in LoadBalancerIngress. If users won't typically care about the difference and we wouldn't provide both, then a single Host field would be better. However, if we would provide both IP and DNS name and/or clients would need to use them differently or in different situations, then two fields would be best.

With respect to the use of "Host" in the API: We use Host in a number of places: HostPort, HostIP, HostNetwork, HostPath, Host (in HTTPGetAction and EventSource), and Hostname. More often than Node (not counting the Node API itself): NodeSelector, NodeName, NodePort. Users have complained about the inconsistency. However, I agree that Hostname makes sense in this case, and is consistent with the other use in the API.